### DISCUSSION TITLE
Hotkey to reset Rp2040

### DISCUSSION THREAD

**Discussion by astralaster** [paulscottrobson/neo6502-firmware#201] [Implemented Ideas]
Currently there is no option to reset the neo6502 other than replugging the power cable. It would be nice to have some kind of hotkey combination like alt GR + r or Ctrl + Alt + del to restart.
Restarting the rp2040 should automatically restart the 6502 as well.

**Comment by jviskari** [comment]
Would it make sense to add some reset function to group 1
- hard reset ( via RP2040 watchdog)
- soft reset (only 65c02, RAM retained, p0=0xff )
- soft reset (65c02, reload RAM from flash. P0 gives the number of 64kB image)

User could map f-key or combination of keys to perform desired action.

**Comment by astralaster** [comment]
Is there already some option to map keys?

**Comment by paulscottrobson** [comment]
Could do both. Could do a hard reset (via WDT) easy enough. It's not that easy to do a soft reset ; from the circuit diagram the we'd have to give up SCL to connect to RESB ?

Affecting the 65C02 doesn't seem to be that practical ; you could hack the read but this seems messy. The code executor that "runs binaries" does it by having a required jmp at the end of the call which either is an RTS or a JSR <code>, without it it won't work. 

In 6502 code it's just JMP ($FFFC) ; theoretically there are some things that might be wrong (decimal mode, interrupt enable etc.) but the first thing you do is set those anyway.

Having a sort of "Ctrl Alt Delete" as a reset, and a reset firmware option is straightforward enough.

**Reply by jviskari** [reply]
Should the RAM image loading from flash should actually be separate feature? Just another command to group 1, which loads image indicated by p0.  Say, the default setting p0=0, would load neobasic. In case of softreset, p0=255 value would just skip loading. There could be values for monitor, cpm65 and perhaps some range for advanced users, who could compile in their own images.

If image has neo header included, loading should be quite straightforward I think.

**Reply by paulscottrobson** [reply]
Possibly. There is a strange bug, which I do not understand as yet. If I try to load BASIC into RAM, a straight copy, with the display running it crashes. I think it's some kind of weird cache/timing bug, because loading files works fine.
